compiler/cli/src/org/jetbrains/kotlin/cli/jklib/JKlibMangler.kt

```diff
@@ -0,0 +1,369 @@
+    append(type.mapToJvmType())
+}
+
+fun KotlinType.mapToJvmType(): JvmType =
+    mapType(this, JvmTypeFactoryImpl, TypeMappingMode.DEFAULT, TypeMappingConfigurationImpl, descriptorTypeWriter = null)
+
+fun hasVoidReturnType(descriptor: CallableDescriptor): Boolean {
+    if (descriptor is ConstructorDescriptor) return true
+    return KotlinBuiltIns.isUnit(descriptor.returnType!!) && !TypeUtils.isNullableType(descriptor.returnType!!)
+            && descriptor !is PropertyGetterDescriptor
+}
+
+fun FunctionDescriptor.computeJvmDescriptor(withReturnType: Boolean = true, withName: Boolean = true): String = buildString {
+    if (withName) {
+        append(if (this@computeJvmDescriptor is ConstructorDescriptor) "<init>" else name.asString())
+    }
+
+    append("(")
+
+    extensionReceiverParameter?.let {
+        appendErasedType(it.type)
+    }
+
+    for (parameter in valueParameters) {
+        appendErasedType(parameter.type)
+    }
+
+    append(")")
+
+    if (withReturnType) {
+        if (hasVoidReturnType(this@computeJvmDescriptor)) {
+            append("V")
+        } else {
+            val returnType = returnType!!
+            // Workaround for b/447548985
+            // Given a class annotated with `@NullMarked`, there will be such difference between K1 and K2 signatures for
+            // functions containing primitive types in arguments/return value:
+            // ```
+            // kotlinjavainterop/NullMarkedClass.getNonNullInteger()I // K2
+            // kotlinjavainterop/NullMarkedClass.getNonNullInteger()Ljava/lang/Integer; // K1
+            // ```
+            // The reason is that in K1 the return type is annotated with `@EnhancedNullability`, but in K2 it isn't.
+            // The workaround forces adding `@EnhancedNullability` for such types so that they are always boxed
+            // when computing function's signature.
+            if (returnType is SimpleType &&
+                KotlinBuiltIns.isPrimitiveType(returnType) &&
+                !returnType.hasEnhancedNullability() &&
+                containingDeclaration.annotations.hasAnnotation(JSPECIFY_NULL_MARKED_ANNOTATION_FQ_NAME)
+            ) {
+                val newType = returnType.replaceAnnotations(
+                    CompositeAnnotations(
+                        returnType.annotations,
+                        ENHANCED_NULLABILITY_ANNOTATIONS
+                    )
+                )
+                appendErasedType(newType)
+            } else {
+                appendErasedType(returnType)
+            }
+        }
+    }
+}
+
+fun CallableDescriptor.computeJvmSignature(): String? = signatures {
+    if (DescriptorUtils.isLocal(this@computeJvmSignature)) return null
+
+    val classDescriptor = containingDeclaration as? ClassDescriptor ?: return null
+    if (classDescriptor.name.isSpecial) return null
+
+    signature(
+        classDescriptor,
+        (original as? FunctionDescriptor ?: return null).computeJvmDescriptor()
```

Comment:
This is the only different line from `org.jetbrains.kotlin.load.kotlin.MethodSignatureMappingKt#computeJvmSignature`. The original:
```
 (original as? SimpleFunctionDescriptor ?: return null).computeJvmDescriptor()
```